{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Backup your exercises\n",
    "copy manually your changed exercise files to another directory\n",
    "\n",
    "## Update your Course Materials!\n",
    "<code>\n",
    "cd cbs-python\n",
    "git checkout -f master\n",
    "git pull origin master\n",
    "\n",
    "\n",
    "In case of \"merge\" problems ...\n",
    "\n",
    "git fetch --all\n",
    "git reset --hard origin/master\n",
    "</code>\n",
    "<br><br><br>\n",
    "<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Last weeks exercises\n",
    "\n",
    "###1. Reading CSV files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "csv_file = \"../data/genes.csv\"\n",
    "\n",
    "fh = open(csv_file)\n",
    "\n",
    "fh.readline() # read but ignore the first line\n",
    "\n",
    "D = {}\n",
    "for line in fh.readlines():\n",
    "        fields = line.split('\\t')\n",
    "        if len(fields) != 3: continue\n",
    "        gene, C1, C2 = fields\n",
    "        D[gene] = (float(C1), float(C2))\n",
    "\n",
    "print D\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The same code using a dictionary of dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "53.3653467074\n"
     ]
    }
   ],
   "source": [
    "\n",
    "csv_file = \"../data/genes.csv\"\n",
    "\n",
    "fh = open(csv_file)\n",
    "\n",
    "fh.readline() # read but ignore the first line\n",
    "\n",
    "D = {}\n",
    "for line in fh.readlines():\n",
    "        fields = line.split('\\t')\n",
    "        if len(fields) != 3: continue\n",
    "        \n",
    "        gene, C1, C2 = fields\n",
    "       \n",
    "        D[gene] = {'C1':float(C1), 'C2':float(C2)}\n",
    "\n",
    "gene = 'RNF14'\n",
    "print D[gene]['C1']\n",
    "\n",
    "fh.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###2. Parsing fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sequence': 'MKVKGTRRNYQHLWRWGTLLLGMLMICSATEKLWVTVYYGVPVWKEATTTLFCASDARAYDTEVHNVWATHACVPTDPNPQEVVLGNVTENFNMWKNNMVEQMQEDIISLWDQSLKPCVKLTPLCVTLNCTDLGKATNTNSSNWKEEIKGEIKNCSFNITTSIRDKIQKENALFRNLDVVPIDNASTTTNYTNYRLIHCNRSVITQACPKVSFEPIPIHYCTPAGFAILKCNNKTFNGKGPCTNVSTVQCTHGIRPIVSTQLLLNGSLAEEEVVIRSDNFTNNAKTIIVQLNESVAINCTRPNNNTRKSIYIGPGRAFHTTGRIIGDIRKAHCNISRAQWNNTLEQIVKKLREQFGNNKTIVFNQSSGGDPEIVMHSFNCRGEFFYCNTTQLFNNTWRLNHTEGTKGNDTIILPCRIKQIINMWQEVGKAMYAPPIGGQISCSSNITGLLLTRDGGTNVTNDTEVFRPGGGDMRDNWRSELYKYKVIKIEPLGIAPTKAKRRVVQREKRAVGIVGAMFLGFLGAAGSTMGAVSLTLTVQARQLLSGIVQQQNNLLRAIEAQQHLLQLTVWGIKQLQARVLAVERYLRDQQLLGIWGCSGKLICTTAVPWNASWSNKSLEDIWDNMTWMQWEREIDNYTNTIYTLLEESQNQQEKNEQELLELDKWASLWNWFSITNWLWYIKIFIMIVGGLVGLRIVFAVLSIVNRVRQGYSPLSFQTRLPVPRGPDRPDGIEEEGGERDRDRSVRLVDGFLALIWEDLRSLCLFSYRRLRDLLLIAARTVEILGHRGWEALKYWWSLLQYWIQELKNSAVSWLNATAIAVTEGTDRVIEVAQRAYRAILHIHRRIRQGLERLLL', 'DB': 'sp', 'AccNum': 'P03378', 'EntryName': 'ENV_HV1A2', 'ProtName': 'Envelope glycoprotein gp160', 'Organism': 'Human immunodeficiency virus type 1 group M subtype B (isolate ARV2/SF2)'}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "fasta_file = \"../data/HIV-1_M-B.fasta\"\n",
    "\n",
    "def parse_fasta(fasta_file):\n",
    "    entries = []\n",
    "    entry = {}\n",
    "\n",
    "    fh = open(fasta_file)\n",
    "    for line in fh.readlines():\n",
    "        if line[0] == '>':\n",
    "            if entry: entries.append(entry)\n",
    "\n",
    "            fields = line.split('|')\n",
    "            db, UniqueIdentifier, description = fields\n",
    "            entry = {}\n",
    "            entry['DB'] = db[1:]\n",
    "            entry[\"AccNum\"] = UniqueIdentifier\n",
    "\n",
    "            d_fields = description.split()\n",
    "\n",
    "# >db|UniqueIdentifier|EntryName ProteinName OS=OrganismName[ GN=GeneName]\n",
    "#  PE=ProteinExistence SV=SequenceVersion\n",
    "\n",
    "#>sp|P03378|ENV_HV1A2 Envelope glycoprotein gp160 OS=Human \\\n",
    "# immunodeficiency virus type 1 group M subtype B (isolate ARV2/SF2) GN=env PE=3 SV=1\n",
    "            protein_part, organism_part = description.split(' OS=')\n",
    "            entry[\"EntryName\"] = protein_part.split()[0]\n",
    "            entry[\"ProtName\"] =  ' '.join(protein_part.split()[1:])\n",
    "            entry[\"Organism\"] = organism_part.split('PE=')[0].split('GN=')[0].strip()\n",
    "            entry['Sequence'] = ''\n",
    "\n",
    "        else:\n",
    "            entry['Sequence'] = entry['Sequence'] + line.strip()\n",
    "        \n",
    "\n",
    "    entries.append(entry)\n",
    "    return entries\n",
    "\n",
    "HIV = parse_fasta(fasta_file)\n",
    "#print 'HIV=', HIV\n",
    "new_entry =  HIV[0]\n",
    "print new_entry"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###3. Writing files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTT F TTC F TTA L TTG L \n",
      "TCT S TCC S TCA S TCG S \n",
      "TAT Y TAC Y TAA * TAG * \n",
      "TGT C TGC C TGA * TGG W \n",
      "CTT L CTC L CTA L CTG L \n",
      "CCT P CCC P CCA P CCG P \n",
      "CAT H CAC H CAA Q CAG Q \n",
      "CGT R CGC R CGA R CGG R \n",
      "ATT I ATC I ATA I ATG M \n",
      "ACT T ACC T ACA T ACG T \n",
      "AAT N AAC N AAA K AAG K \n",
      "AGT S AGC S AGA R AGG R \n",
      "GTT V GTC V GTA V GTG V \n",
      "GCT A GCC A GCA A GCG A \n",
      "GAT D GAC D GAA E GAG E \n",
      "GGT G GGC G GGA G GGG G \n",
      "\n"
     ]
    }
   ],
   "source": [
    "def write_codon_table(filename, codon_table):\n",
    "    fh = open(filename, 'w+')\n",
    "    for base1 in bases:\n",
    "        for base2 in bases:\n",
    "            for base3 in bases:\n",
    "                codon = base1+base2+base3\n",
    "                fh.write('%s %s ' % (codon, codon_table[codon]))\n",
    "            fh.write('\\n')\n",
    "    fh.close()\n",
    "\n",
    "bases = ['T', 'C', 'A', 'G']\n",
    "codons = [a+b+c for a in bases for b in bases for c in bases]\n",
    "amino_acids = 'FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG'\n",
    "codon_table = dict(zip(codons, amino_acids))\n",
    "\n",
    "write_codon_table('mycodon_table.txt', codon_table)\n",
    "\n",
    "# quickly checking what we have written\n",
    "print open('mycodon_table.txt').read()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###4. Sequence sorter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fh_bacteria = open('../data/eftu_bacteria.faa', 'w')\n",
    "fh_archaea = open('../data/eftu_archaea.faa', 'w')\n",
    "fh_eukaryota = open('../data/eftu_eukrayota.faa', 'w')\n",
    "\n",
    "for line in open('../data/eftu.faa'):\n",
    "    if line[0] == '>':\n",
    "        kingdom = line.split()[1]\n",
    "        if kingdom == 'Bacteria;':\n",
    "            out_stream = fh_bacteria\n",
    "        elif kingdom == 'Archaea;':\n",
    "            out_stream = fh_archaea\n",
    "        elif kingdom == 'Eukaryota;':\n",
    "            out_stream = fh_eukaryota\n",
    "\n",
    "    out_stream.write(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###5. DNA translator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fasta_file = '../data/human_beta_globin.fasta'\n",
    "\n",
    "def read_fasta(fastafile):\n",
    "    fh = open(fastafile)\n",
    "    entries = {}\n",
    "    for line in fh.readlines():\n",
    "        if line[0]  == '>':\n",
    "            seq_name = line[1:].strip()\n",
    "            entries.setdefault(seq_name,'')\n",
    "        else:\n",
    "            entries[seq_name] += line.strip().upper()\n",
    "\n",
    "    return entries\n",
    "\n",
    "def translate(sequence, frame=1):\n",
    "    bases = ['T', 'C', 'A', 'G']\n",
    "    codons = [a+b+c for a in bases for b in bases for c in bases]\n",
    "    amino_acids = 'FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG'\n",
    "    codon_table = dict(zip(codons, amino_acids))\n",
    "    aa_seq = []\n",
    "    for i in range(frame-1, len(sequence), 3):\n",
    "        codon = sequence[i:i+3]\n",
    "        if len(codon) !=3: break\n",
    "        aa_seq.append(codon_table.get(codon, 'X'))\n",
    "    return ''.join(aa_seq)\n",
    "\n",
    "def reverse_complement(sequence):\n",
    "    rseq = sequence[::-1]\n",
    "    rev_dict = {'A':'T','T':'A','G':'C','C':'G'}\n",
    "    return ''.join([rev_dict.get(x, 'N') for x in rseq])\n",
    "                   \n",
    "            \n",
    "for name, sequence in read_fasta(fasta_file).items():\n",
    "    print name\n",
    "    print sequence\n",
    "    for i in range(1,4):\n",
    "        print i, translate(sequence, frame=i)\n",
    "\n",
    "    r_sequence = reverse_complement(sequence)\n",
    "    for i in range(1,4):\n",
    "        print i*-1, translate(r_sequence, frame=i)\n",
    "    \n",
    "\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br><br><br><br><br><br><br><br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "    div.cell{\n",
       "        width:800px;\n",
       "        margin-left:16% !important;\n",
       "        margin-right:auto;\n",
       "    }\n",
       "    div.text_cell_render{\n",
       "        font-family: \"Helvetica Neue\", Arial, Helvetica, Geneva, sans-serif;\n",
       "        line-height: 145%;\n",
       "        font-size: 100%;\n",
       "        width:800px;\n",
       "        margin-left:auto;\n",
       "        margin-right:auto;\n",
       "    }\n",
       "    .CodeMirror{\n",
       "            font-family: \"Source Code Pro\", source-code-pro,Consolas, monospace;\n",
       "    }\n",
       "    div.CodeMirror{\n",
       "    \tfont-size: 11pt;\n",
       "    }\n",
       "    div.cell.border-box-sizing.code_cell{\n",
       "    \tpadding-left: 50px;\n",
       "    }\n",
       "    div.text_cell_input {\n",
       "    \tmargin-left: 50px;\n",
       "    }\n",
       "    .prompt{\n",
       "        display: None;\n",
       "    }\n",
       "    .text_cell_render h1 {\n",
       "        font-family: Helvetica, serif;\n",
       "        font-size: 30pt;\n",
       "        color: #274257;\n",
       "        text-transform: uppercase;\n",
       "        line-height: 100%;\n",
       "    }\n",
       "    .text_cell_render h2 {\n",
       "    \tmargin-top: 35px;\n",
       "        font-family: Helvetica, serif;\n",
       "        font-size: 24pt;\n",
       "        color: #2A75A9;\n",
       "        line-height: 100%;\n",
       "    }\n",
       "    .text_cell_render h3{\n",
       "    \tmargin-top: 35px;\n",
       "        font-family: Helvetica, serif;\n",
       "        font-size: 18pt;\n",
       "        color: #7EB5D6;\n",
       "        line-height: 100%;\n",
       "       }\n",
       "    .text_cell_render h4{\n",
       "    \tmargin-top: 35px;\n",
       "    \tfont-family: Helvetica, serif;\n",
       "        font-size: 14pt;\n",
       "        line-height: 100%;\n",
       "    }\n",
       "    .text_cell_render h5 {\n",
       "        font-weight: 300;\n",
       "        font-size: 22pt;\n",
       "        color: #4057A1;\n",
       "        font-style: italic;\n",
       "        margin-bottom: .5em;\n",
       "        margin-top: 0.5em;\n",
       "        display: block;\n",
       "        line-height: 100%;\n",
       "    }\n",
       "    \n",
       "    .warning{\n",
       "        color: rgb( 240, 20, 20 )\n",
       "        } \n",
       "    .wds {\n",
       "        background: none repeat scroll 0% 0% rgb(247, 247, 247);\n",
       "        padding: 0em 0em;\n",
       "        font-size: 11px;\n",
       "        font-family: Arial,Helvetica,sans-serif;\n",
       "        display: inline-block;\n",
       "        margin: 0px 0em;\n",
       "        white-space: nowrap;\n",
       "    }\n",
       "    cb {\n",
       "        background: none repeat scroll 0% 0% rgb(247, 247, 247);\n",
       "        padding: 0.1em 0.6em;\n",
       "        border: 1px solid rgb(204, 204, 204);\n",
       "        font-size: 11pt;\n",
       "        font-family: \"Source Code Pro\", source-code-pro,Consolas, monospace;\n",
       "        color: rgb(51, 51, 51);\n",
       "        box-shadow: 0px 1px 0px rgba(0, 0, 0, 0.2), 0px 0px 0px 2px rgb(255, 255, 255) inset;\n",
       "        border-radius: 3px 3px 3px 3px;\n",
       "        display: inline-block;\n",
       "        margin: 0px 0.1em;\n",
       "        line-height: 1.4;\n",
       "        white-space: nowrap;\n",
       "    }\n",
       "    q {\t\n",
       "        font-weight: 300;\n",
       "        font-size: 14pt;\n",
       "        color: #4057A1;\n",
       "        margin-bottom: .5em;\n",
       "        margin-top: 0.5em;\n",
       "        display: block;\n",
       "        font-family: \"Helvetica Neue\", Arial, Helvetica, Geneva, sans-serif;\n",
       "        line-height: 145%;\n",
       "        width:800px;\n",
       "        margin-left:auto;\n",
       "        margin-right:auto;\n",
       "    }\n",
       "    qq {\t\n",
       "        font-weight: 300;\n",
       "        font-size: 11pt;\n",
       "        color: #4057A1;\n",
       "        margin-bottom: .5em;\n",
       "        margin-top: 0.5em;\n",
       "        display: block;\n",
       "        font-family: \"Helvetica Neue\", Arial, Helvetica, Geneva, sans-serif;\n",
       "        line-height: 145%;\n",
       "        width:800px;\n",
       "        margin-left:auto;\n",
       "        margin-right:auto;\n",
       "    }\n",
       "    kbd {\n",
       "        background: none repeat scroll 0% 0% rgb(247, 247, 247);\n",
       "        padding: 0.1em 0.6em;\n",
       "        border: 1px solid rgb(204, 204, 204);\n",
       "        font-size: 11px;\n",
       "        font-family: Arial,Helvetica,sans-serif;\n",
       "        color: rgb(51, 51, 51);\n",
       "        box-shadow: 0px 1px 0px rgba(0, 0, 0, 0.2), 0px 0px 0px 2px rgb(255, 255, 255) inset;\n",
       "        border-radius: 3px 3px 3px 3px;\n",
       "        display: inline-block;\n",
       "        margin: 0px 0.1em;\n",
       "        text-shadow: 0px 1px 0px rgb(255, 255, 255);\n",
       "        line-height: 1.4;\n",
       "        white-space: nowrap;\n",
       "    }\n",
       "    ol {\n",
       "        list-style-type: decimal;\n",
       "    }\n",
       "\n",
       "\t.box_rotate {\n",
       "\t\t  -webkit-transform: rotate(270deg);  /* Chrome, Safari 3.1+ */\n",
       "\t\t     -moz-transform: rotate(270deg);  /* Firefox 3.5-15 */\n",
       "\t\t      -ms-transform: rotate(270deg);  /* IE 9 */\n",
       "\t\t       -o-transform: rotate(270deg);  /* Opera 10.50-12.00 */\n",
       "\t\t          transform: rotate(270deg);  /* Firefox 16+, IE 10+, Opera 12.10+ */\n",
       "\t}\n",
       "\n",
       "\t.vertical-text {\n",
       "\t\tdisplay: inline-block;\n",
       "\t\toverflow: hidden;\n",
       "\t\twidth: 1.5em;\n",
       "\t}\n",
       "\n",
       "\t.vertical-text__inner {\n",
       "\t\tdisplay: inline-block;\n",
       "\t\twhite-space: nowrap;\n",
       "\t\tline-height: 1.5;\n",
       "\t\t  -webkit-transform: rotate(270deg);  /* Chrome, Safari 3.1+ */\n",
       "\t\t     -moz-transform: rotate(270deg);  /* Firefox 3.5-15 */\n",
       "\t\t      -ms-transform: rotate(270deg);  /* IE 9 */\n",
       "\t\t       -o-transform: rotate(270deg);  /* Opera 10.50-12.00 */\n",
       "\t\t          transform: rotate(270deg);  /* Firefox 16+, IE 10+, Opera 12.10+ */\n",
       "\t}\n",
       "\n",
       "\t.vertical-text__inner:after {\n",
       "\t\tcontent: \"\";\n",
       "\t\tdisplay: block;\n",
       "\t\tmargin: -1.5em 0 100%;\n",
       "\t}\n",
       "\n",
       "\t.topics{\n",
       "\t\tbackground: url('files/images/topics.png') no-repeat scroll 0% 0%; \n",
       "\t\twidth: 520px; \n",
       "\t\theight: 250px; \n",
       "\n",
       "\t\tpadding-top: 35px;\n",
       "\t}\n",
       "\n",
       "\t.c_program{\n",
       "\t\tbackground: url('files/images/program.png') no-repeat scroll 0% 0%; width: 520px; margin-left: auto; margin-right: auto;\n",
       "\t}\n",
       "\n",
       "\t.c_require{\n",
       "\t\tbackground: url('files/images/software.png') no-repeat scroll 0% 0%; width: 520px; height: 250px; margin-left: auto; margin-right: auto;\n",
       "\t}\n",
       "\n",
       "</style>\n",
       "<script>\n",
       "    MathJax.Hub.Config({\n",
       "                        TeX: {\n",
       "                           extensions: [\"AMSmath.js\"]\n",
       "                           },\n",
       "                tex2jax: {\n",
       "                    inlineMath: [ ['$','$'], [\"\\\\(\",\"\\\\)\"] ],\n",
       "                    displayMath: [ ['$$','$$'], [\"\\\\[\",\"\\\\]\"] ]\n",
       "                },\n",
       "                displayAlign: 'center', // Change this to 'center' to center equations.\n",
       "                \"HTML-CSS\": {\n",
       "                    styles: {'.MathJax_Display': {\"margin\": 4}}\n",
       "                }\n",
       "        });\n",
       "</script>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML at 0x10c7b40d0>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.core.display import HTML\n",
    "\n",
    "\n",
    "def css_styling():\n",
    "    styles = open(\"../styles/custom.css\", \"r\").read()\n",
    "    return HTML(styles)\n",
    "css_styling()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
