{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 03 - Exercises ..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Decision Statements\n",
    "\n",
    "Explain what is going on in this code block:\n",
    "\n",
    "    if conditionA or conditionB:\n",
    "        \" code A \"\n",
    "    else:\n",
    "        \" code B \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Playground: Use this to test some code if you like.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, explain what is going on in this code block:\n",
    "    \n",
    "    if conditionA:\n",
    "        \" code A \"\n",
    "    elif conditionB:\n",
    "        \" code B \"\n",
    "    else:\n",
    "        \" code C \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Playground: Use this to test some code if you like.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Net charge of a Protein\n",
    "\n",
    "Dictionary containing charges of individual amino acids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "AACharge = {\"C\": -0.045,\n",
    "            \"D\": -0.999,\n",
    "            \"E\": -0.998,\n",
    "            \"H\": 0.091,\n",
    "            \"K\": 1.0,\n",
    "            \"R\": 1.0,\n",
    "            \"Y\": -0.001,}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make a function ```prot_charge``` that takes a protein sequence as input and calculates the net charge using the dictionary above\n",
    "\n",
    "**Instructions**\n",
    "<ol>\n",
    "<li>Define the function ```prot_charge``` to accept a protein sequence.</li>\n",
    "<li>Validate that the sequence is infact a protein.</li>\n",
    "<li>Create a variable to hold the protein charge. Set it to -0.002 initially, since this is the charge of the free ends.</li>\n",
    "<li>Compute the charge of the full protein sequence and add it to this variable</li>\n",
    "<li>Return this charge variable</li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function should not take into account:\n",
    "<ul><li>surface accesibility, pH, cysteine disulfide bridges or amidation and acetylation.</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.046\n"
     ]
    }
   ],
   "source": [
    "# Make the function here\n",
    "def prot_charge(seq):\n",
    "    valid_aa = 'ACEDGFIHK*MLNQPSRTWVY'\n",
    "\n",
    "    if not all([(x in valid_aa) for x in seq.upper()]):\n",
    "        print \"not a protein sequence\"\n",
    "        return -1\n",
    "    \n",
    "    charge = 0\n",
    "\n",
    "    for aa in seq:\n",
    "        charge += AACharge.get(aa,0)\n",
    "        \n",
    "# or usin list comprehension and the sum function\n",
    "#    charge = sum([AACharge.get(x,0) for x in seq])\n",
    "    return charge\n",
    "\n",
    "\n",
    "print prot_charge(\"MCIYLTEERRISHMFPKSPNRQMYPQPGQQPYTPYPIPQLPPMAQK\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Printing the DNA codon table using *nested loops*\n",
    "In this exercise we will be using a dictionary containing the <a href=\"http://en.wikipedia.org/wiki/DNA_codon_table\">DNA codon table</a>. Instead of generating this dictionary manually by typing in the **three nucleotides as key** and an **amino acid as value**, like\n",
    "\n",
    "    codon_table[\"AAA\"] = \"K\"\n",
    "    codon_table[\"AAC\"] = \"N\"\n",
    "    ...\n",
    "\n",
    "we will use a shortcut:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Hint: Make sure to run me! \n",
    "# also, try to understand what is happening here\n",
    "# new is the zip function, look it up with \"zip?\"\n",
    "\n",
    "bases = ['T', 'C', 'A', 'G']\n",
    "codons = [a+b+c for a in bases for b in bases for c in bases]\n",
    "amino_acids = 'FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG'\n",
    "codon_table = dict(zip(codons, amino_acids))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Which gives the same resulting dictionary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print codon_table[\"AAA\"]\n",
    "print codon_table[\"AAC\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make a function ```print_codon_table()``` that prints the codon table in a nice, but simple arrangement, using nested for loops like below. Use the dictionary ```codon_table``` as well as the ```bases``` variable:\n",
    "\n",
    "    for base1 in bases:\n",
    "        for base2 in bases:\n",
    "            for base3 in bases:\n",
    "                print ...\n",
    "\n",
    "We have already begun the function for you."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAA K AAT N AAC N AAG K\n",
      "ATA I ATT I ATC I ATG M\n",
      "ACA T ACT T ACC T ACG T\n",
      "AGA R AGT S AGC S AGG R\n",
      "TAA * TAT Y TAC Y TAG *\n",
      "TTA L TTT F TTC F TTG L\n",
      "TCA S TCT S TCC S TCG S\n",
      "TGA * TGT C TGC C TGG W\n",
      "CAA Q CAT H CAC H CAG Q\n",
      "CTA L CTT L CTC L CTG L\n",
      "CCA P CCT P CCC P CCG P\n",
      "CGA R CGT R CGC R CGG R\n",
      "GAA E GAT D GAC D GAG E\n",
      "GTA V GTT V GTC V GTG V\n",
      "GCA A GCT A GCC A GCG A\n",
      "GGA G GGT G GGC G GGG G\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def print_codon_table():\n",
    "    for base1 in bases:\n",
    "        for base2 in bases:\n",
    "            for base3 in bases:\n",
    "                codon = base1+base2+base3\n",
    "                print codon, codon_table[codon],\n",
    "                # OR\n",
    "                #print \"%s %s\\t\" % (codon, codon_table[codon]),\n",
    "            print\n",
    "            \n",
    "\n",
    "print_codon_table()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Hint:** *Use *\n",
    "\n",
    "    print \"\"\n",
    "    \n",
    "*to make a new line in each nested loop.*\n",
    "\n",
    "Test the function here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print_codon_table()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See if you can arrange the output like this: "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "TTT F   TTC F   TTA L   TTG L   \n",
    "TCT S   TCC S   TCA S   TCG S   \n",
    "TAT Y   TAC Y   TAA *   TAG *   \n",
    "TGT C   TGC C   TGA *   TGG W   \n",
    "\n",
    "CTT L   CTC L   CTA L   CTG L   \n",
    "CCT P   CCC P   CCA P   CCG P   \n",
    "CAT H   CAC H   CAA Q   CAG Q   \n",
    "CGT R   CGC R   CGA R   CGG R   \n",
    "\n",
    "ATT I   ATC I   ATA I   ATG M   \n",
    "ACT T   ACC T   ACA T   ACG T   \n",
    "AAT N   AAC N   AAA K   AAG K   \n",
    "AGT S   AGC S   AGA R   AGG R   \n",
    "\n",
    "GTT V   GTC V   GTA V   GTG V   \n",
    "GCT A   GCC A   GCA A   GCG A   \n",
    "GAT D   GAC D   GAA E   GAG E   \n",
    "GGT G   GGC G   GGA G   GGG G"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Calculating the coding potential of a DNA sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unequal usage of codons in the coding regions appears to be a universal feature of the genomes across the phylogenetic spectra. This bias obeys mainly to\n",
    "<ul><li>the uneven usage of the amino acids in the existing proteins, and</li>  <li> the uneven usage of synonymous codons</li></ul>\n",
    "The bias in the usage of the synonymous codons correlates with the abundance of the corresponding tRNAs. The correlation is particularly strong for highly expressed genes. Codon usage is specific of the taxonomic group, and there exist correlation between taxonomic divergence and similarity of codon usage. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By comparing the frequency of codons in a region of an species genome read in a given frame with the typical frequency of codons in the species genes, it is possible to estimate a likelihood of the region coding for a protein in such a frame. Regions in which codons are used with frequencies similar to the typical species codon frequencies are likely to code for genes (**exons**) while regions which codons distribution is uniform -- 1/64 (0.015625) -- could be considered as non-coding regions (**introns**). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center>\n",
    "<table>\n",
    "<tr><td>\n",
    "<font SIZE=\"-2\"> </font>\n",
    "<table CELLPADDING=3 BORDER=\"1\">\n",
    "<tr><td ALIGN=\"CENTER\" COLSPAN=16>The Human Codon Usage Table</td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Gly </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GGG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 17.08 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.23  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Arg </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> AGG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 12.09 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.22  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Trp </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TGG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 14.74 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 1.00  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Arg </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CGG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 10.40 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.19 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Gly </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GGA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 19.31 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.26  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Arg </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> AGA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 11.73 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.21  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> End </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TGA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 2.64 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.61  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Arg </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CGA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 5.63 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.10 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Gly </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GGT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 13.66 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.18  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ser </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> AGT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 10.18 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.14  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Cys </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TGT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 9.99 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.42  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Arg </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CGT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 5.16 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.09 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Gly </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GGC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 24.94 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.33  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ser </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> AGC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 18.54 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.25  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Cys </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TGC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 13.86 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.58  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Arg </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CGC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 10.82 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.19 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Glu </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GAG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 38.82 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.59  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Lys </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> AAG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 33.79 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.60  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> End </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TAG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.73 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.17  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Gln </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CAG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 32.95 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.73 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Glu </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GAA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 27.51 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.41  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Lys </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> AAA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 22.32 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.40  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> End </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TAA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.95 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.22  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Gln </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CAA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 11.94 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.27 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Asp </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GAT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 21.45 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.44  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Asn </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> AAT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 16.43 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.44  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Tyr </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TAT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 11.80 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.42  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> His </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CAT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 9.56 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.41 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Asp </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GAC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 27.06 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.56  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Asn </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> AAC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 21.30 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.56  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Tyr </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TAC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 16.48 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.58  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> His </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CAC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 14.00 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.59 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Val </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GTG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 28.60 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.48  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Met </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> ATG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 21.86 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 1.00  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Leu </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TTG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 11.43 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.12  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Leu </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CTG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 39.93 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.43 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Val </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GTA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 6.09 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.10  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ile </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> ATA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 6.05 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.14  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Leu </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TTA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 5.55 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.06  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Leu </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CTA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 6.42 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.07 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Val </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GTT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 10.30 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.17  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ile </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> ATT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 15.03 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.35  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Phe </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TTT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 15.36 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.43  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Leu </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CTT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 11.24 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.12 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Val </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GTC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 15.01 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.25  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ile </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> ATC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 22.47 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.52  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Phe </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TTC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 20.72 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.57  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Leu </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CTC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 19.14 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.20 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ala </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GCG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 7.27 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.10  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Thr </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> ACG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 6.80 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.12  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ser </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TCG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 4.38 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.06  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Pro </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CCG </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 7.02 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.11 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ala </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GCA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 15.50 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.22  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Thr </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> ACA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 15.04 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.27  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ser </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TCA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 10.96 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.15  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Pro </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CCA </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 17.11 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.27 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ala </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GCT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 20.23 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.28  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Thr </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> ACT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 13.24 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.23  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ser </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TCT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 13.51 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.18  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Pro </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CCT </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 18.03 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.29 </font></td>\n",
    "</tr>\n",
    "<tr><td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ala </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> GCC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 28.43 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.40  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Thr </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> ACC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 21.52 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.38  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Ser </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> TCC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 17.37 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.23  </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> Pro </font></td>\n",
    "<td ALIGN=\"LEFT\"><font SIZE=\"-2\"> CCC </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 20.51 </font></td>\n",
    "<td ALIGN=\"RIGHT\"><font SIZE=\"-2\"> 0.33 </font></td>\n",
    "</tr>\n",
    "<tr><td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "<td></td>\n",
    "</tr>\n",
    "<caption>\n",
    "<font size=1>\n",
    "<strong>Table 1:</strong>\n",
    "The human codon usage and codon preference table as published in \n",
    "<tt> <a href=\"http://bioinformatics.weizmann.ac.il/databases/codon\">Weizmann Institute of Science</a></tt>. For each \n",
    "codon, the table displays the frequency of usage of each codon (per thousand)\n",
    "in human coding regions (first column) and the relative frequency of each\n",
    "codon among synonymous codons (second column).\n",
    "</font>\n",
    "</caption>\n",
    "</table>\n",
    "</table>\n",
    "</center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let $S$ be a sequence of DNA, the higher the number of codons in $S$ matching the more frequent codons in the table, the *higher the coding potential* of the sequence $S$ (the probability to be encoding a protein). This value can be computed as: \n",
    "\n",
    "$$ \\frac{p(C_1)p(C_2)p(C_3)...p(C_m)}{\\frac{1}{64}\\frac{1}{64}\\frac{1}{64}...\\frac{1}{64}} $$\n",
    "\n",
    "where $p(C_1),p(C_2),p(C_3)...p(C_m)$ are the frequency of codons as $C_1=\\{ACC \\}$. \n",
    "\n",
    "This expression is known as *likelihood ratio* and it is usually computed in logarithmics terms to reduce the magnitude of the numbers (*log-likelihood ratio*). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file *propcodons.txt* in your data folder contains the human codon usage table above in a format like"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "GGG 0.01708\n",
    "GGA 0.01931\n",
    "GGT 0.01366\n",
    "GGC 0.02494\n",
    "GAG 0.03882\n",
    "GAA 0.02751\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The code snippet below reads this file and creates a dictionary (```codonprop```) containing the frequencies. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Loads a file containing codon frequencies\n",
    "# into a dictionary called codonprop\n",
    "codonprop = {}                               # Initiates an empty dictionary\n",
    "with open(\"../data/propcodons.txt\") as f:    # Opens the file\n",
    "    line = f.readline()                      # Reads the very first line\n",
    "    while line:                              # Reads the rest of the lines uptil it comes up empty.\n",
    "        line = line.split()                  # Converts line into a list of text separated by spaces\n",
    "        codonprop[line[0]] = float(line[1])  # Puts the value into a dictionary with codon as key\n",
    "        line = f.readline()                  # Reads the next line"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:** *We learn more about dealing with files next week.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can access the frequencies using the codon you are looking for, like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.01324\n"
     ]
    }
   ],
   "source": [
    "print codonprop.get(\"ACT\", 1.0/64)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make a function ```coding_potential(seq)``` that computes the log-likelihood value of a sequence by following the instructions below"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<ol>\n",
    "<li>Go through the sequence in steps of three using an iterator $i$</li>\n",
    "    *Hint: Use the function ```range(start, stop[, step])``` as the iteration.*\n",
    "<li>Read the $i^{th}$ codon into a variable using slicing.</li>\n",
    "<li>Compute the log likelihood of the sequence using $r = r + log(p_i) - log(1.0/64)$.</li>\n",
    "<li>Return the log likelihood value</li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from math import log\n",
    "\n",
    "def coding_potential(seq):\n",
    "    \"\"\" Returns the coding potential of a DNA sequence \"\"\"\n",
    "    r = 0\n",
    "    L = len(seq)\n",
    "    for i in range(0, L, 3):\n",
    "        codon = seq[i:i+3]\n",
    "        pi = codonprop.get(codon, 1.0/64)\n",
    "        r=r+log(pi) - log(1.0/64)\n",
    "    return r\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To test the program we will use it on two different sequences, an exon and an intron:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Make sure to run me! \n",
    "exon = \"GCTGCTGGTGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACT\\\n",
    "CCTGATGCTGTTATGGGCAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCT\\\n",
    "TTAGTGATGGCCTGGCTCACCTGGACAACCTCAAGGGCACCTTTGCCACACTGAGTGAGCT\\\n",
    "GCACTGTGACAAGCTGCACGTGGATCCTGAGAACTTCAGG\"\n",
    "\n",
    "intron = \"ATAACAATTGTTTTCTTTTGTTTAATTCTTGCTTTCTTTTTTTTTCTTCTCCGCAATTTTT\\\n",
    "ACTATTATACTTAATGCCTTAACATTGTGTATAACAAAAGGAAATATCTCTGAGATACATT\\\n",
    "AAGTAACTTAAAAAAAAACTTTACACAGTCTGCCTAGTACATTACTATTTGGAATATATGT\\\n",
    "GTGCTTATTTGCATATTCATAATCTCCCTACTTTATTTTC\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Call the functions using these sequences by running: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-16.131510449320476"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_potential(exon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-14.364827936642822"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_potential(intron)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function should return approximately: \n",
    "    $$r_{exon} = -16.1315104493$$\n",
    "    $$r_{intron} = -14.3648279366$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unfortunately, as you have seen, the log-likelihood value computed for both cases is very similar so that we can not distinguish among coding and non coding regions. As you know, a DNA sequence can be translated into a series of amino acids using three different reading frames and we are probably using the wrong frame right now. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Copy your function ```coding_potential(seq)``` into the code block below and update it to compute the maximum coding potential of all three reading frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def coding_potential2(seq):\n",
    "    \"\"\" Returns the maximal coding potential of a DNA sequence \"\"\"\n",
    "    R = []\n",
    "    for i in range(3):\n",
    "        R.append(coding_potential(seq[i:]))\n",
    "    return max(R)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Call the functions using these sequences by running: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24.064576314294488"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_potential2(exon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-14.364827936642822"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_potential2(intron)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now your function should return approximately:\n",
    "    $$r_{exon} = 24.0645763143$$\n",
    "    $$r_{intron} = -14.3648279366$$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In practice, the problem is not to determine the likelihood that a given sequence is coding or not, but to locate the (usually small) coding regions within large genomic sequences. The typical procedure is to compute the value of a coding statistic in successive (overlapped) sliding windows, and record the value of the statistic for each of the windows. This generates a profile along the sequence in which peaks may point to the coding regions and valleys to the non-coding ones. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a function ```exon_predictor(seq, window, overlap)``` that computes the protein coding potential using sliding windows in the three reading frames at the same time. Follow these instructions:\n",
    "\n",
    "<ol>\n",
    "<li>Initialize a list $r$ to be filled up with log-likelihood ratios</li>\n",
    "<li>Make a loop that goes through all windows in the sequence shifted by the ```overlap``` size</li>\n",
    "*Hint: Use ```range(start, stop[, step])```*\n",
    "<li>Create a segment sequence corresponding to the ```window``` size using slicing (```[i:i+window]```)</li>\n",
    "<li>Compute the maximum coding potential in this segment using ```coding_potential(seq)``` and append it to $r$</li>\n",
    "<li>Return $r$</li>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def exon_predictor(seq, window=120, overlap=10):\n",
    "    \"\"\" Returns the the log-likelihood ratio computed for \n",
    "        every [window] nucleotide windows, shifted by [overlap]  \n",
    "        nucleotides in the sequence seq.\"\"\"\n",
    "\n",
    "    R = []\n",
    "    for i in range(0, len(seq), overlap):\n",
    "        s = seq[i:i+window]\n",
    "        R.append(coding_potential2(s))\n",
    "\n",
    "    return R\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To test our prediction tool we will use the sequence that encodes three exons of the Human $\\beta$-globin gene. To analyze this information, we can visualize the results in a graphical way using the program matplotlib. To do this, run the code below: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n",
      "(Ignore any warnings...)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: pylab import has clobbered these variables: ['record', 'log', 'f']\n",
      "`%matplotlib` prevents importing * from pylab and numpy\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Container object of 206 artists>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEACAYAAAC6d6FnAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFNtJREFUeJzt3W/MLGdZx/HfDw684E9SG0xbSk1rBAWDthALCZDuG+vp\nG0pfIPBCiRhDgqJRjFBM3D3xjZCUqDESIi0p/imiBqRBtAfDYhODDVLaA+VYTuhJeqCcEilKExNr\ne/nimeGZM53dZ3dm7vn7/SRPzs7+m5n73DvXXNd9z64jQgCA+XlG3xsAAOgHAQAAZooAAAAzRQAA\ngJkiAADATBEAAGCmGgcA27fZPm/7VOG+le1ztu/N/o43XQ8AoF1tZAAfkVQ+wIekD0TENdnfP7aw\nHgBAixoHgIi4W9JjFQ+56XsDANJJOQbwTtv32b7V9kUJ1wMAqCFVAPigpKskXS3pEUm3JFoPAKCm\nYyneNCIezW/b/rCkO8vPsc2XEAFADRHRSok9SQZg+7LC4k2STlU9LyJm97dcLne6b65/tAVtQVts\n/2tTG9NA75D0r5J+3PbDtt8m6X2277d9n6TrJP1m0/VMxYkTJ/reBACQ1M4soLdExAsj4tkRcUVE\n3BYRvxgRPxURPx0Rb4iI821sLDBFq9Wq703ATHEl8AAsFou+N2Ew5tgWm7LCObbFJrRFGm67prTz\niu3oa919st16HQ/jtFqttFqt6BPYS9ZfWhkEJgB0jA87cnlfoE9gH20GAEpAADBTBIAe2JbNN2UA\n6BcBoAPM8gAwRASADuSzPAgEAIaEANCh8nQ/AgKAPjELqAPF2R5lc2kDPB2zgFAHs4CAkSP7wxCQ\nAXSADABleV8gA8C+yAAAAI0RABIj1QcwVASAhFarFV//DGCwGANIqFjzZwwARYwBoC7GAAAAjREA\nAGCmCAAAMFMEgASY+QNgDBgETqDqwi8GgVHEIDDqYhB4wDj7x1HoIxgKMoCWlc/scmQAyFX1C/oB\ndkUGMCGcDQLoCxlAy/bNAPLHMB9kAGiCDGBAOIMHMFZkAA2Vz97IAHAUMgA0QQYAAGiMAAAAM9U4\nANi+zfZ526cK911s+6TtB23fZfuipusBALSrjQzgI5KOl+57j6STEfESSf+cLQMABqRxAIiIuyU9\nVrr79ZJuz27fLukNTdczZE1nAtneOEAMAKm0MgvI9pWS7oyIl2fLj0XED2W3Lem7+XLhNZOZBSQd\nzuSpMwuo+BxMH7OA0MSoZgFlR/lZ9G6uCQAwJscSve9525dGxLdtXybp0aonFQ+Yi8VCi8Wi1spW\nq9UgDr78/i+2GUo/xbis12ut1+sk752qBPR+Sf8ZEe+z/R5JF0XEe0qvaa0E1FcKXfzR96NKPpSA\nsO1rwpfLJcEBO2mzBNQ4ANi+Q9J1kl4g6byk35P095I+LulHJJ2V9PMR8b3S60YfAPY54BMAcNTv\nRPD/j10MKgDUXjEB4Gk4AEwbAQBtGNUgMABgmCYXAKijAsBuJlcC6rIcRAkI+6AEhDZQAuoZWQaA\nKSADaLA+MgDsgwwAbSAD2IAzcwDY3aQygK7OpMgAUAcZANpABgAAaGz0AaCq7EMpCACONvoS0Lav\nX06FEhDqoASENlACAiaIzBVdIwOouU4yAOxr134DbEMGkJnaGdPU9gfAsI06A+jrTCpVBpA/D9NE\nBoA2kAEAABojAADATBEAAGCmCAAAMFMEAACYKQLAwDAVFEBXmAZac72ppoHmz8X0MA0UbWAaKACg\nMQIAAMwUAQAAZmryAYBBVQCoNvkAcOLEidqvJXigD/Q7dGXys4DymRd137/82tVqpRMnTjALCHvb\nZ/YYfQCbMAuoR00yCswXZ/UYokkHgPxDx4cPfePEAUOUNADYPmv7ftv32r4n5bqq5B+6tsYBCCQA\npiTpGIDthyS9MiK+W/FY8jGApldYln+su0nNnzGAedv206VVy8vlkhMOVBrbGEArG9ql8gePDyK6\nRskIXUidAXxD0n9JelLShyLizwqPDTYDaOO7fsgAULRvBpDfBsrazACOtfEmW7wmIh6x/cOSTto+\nHRF35w8Wz6wXi4UWi0XizQGAcVmv11qv10neu7PrAGwvJT0eEbdky2QAG3DmNz1kAGjLKMYAbD/H\n9vOz28+VdL2kU6nWBwDYT8oS0CWSPpGdzRyT9JcRcVfC9QEA9pAsAETEQ5KuTvX+AIBmJn0lMABg\nMwIAAMwUAaCEi77QFvoShm6UASDlB4srMNGWvC8RCDBUowsA+ffxA2PRtL8SQJDK6H4Qps4FNfly\n6vfnQjAUNb2gsI0fNcL0jOJCsCHiTAoADs0qA9jlTIoMAG0hA0AKZAAAgMYIAMBAUbJEapSAWn7/\nXZd3Qdo/bm1+qyx9ATlKQACAxkYVAEiJAaA9vQaAfUsibVwARhABgAO9jgFI+/9SV66N6XXb1sEY\nAJpiDAApMAYAAGiMAAAAM0UAKGB8AMCcMAaQ4P13Wd4Fdd9xYwwAKTAGAICMFY0RAIAR4ncx0Ibe\nA0B+FmNbtivPajjTAS7EwR9t6H0MQDr6x1varKVK0nK5rAwqjAGgTSnHAHa5pgXTxBhAQ5w9AfWR\nkU/HoANAVx1tqB16qNuFeeMEajoGXQIqptApSjKFbUle8qlTAipvJ8ZlqiUgpqX2a3IlIM50gaMN\n8XMyxG3C7gaRAZSRARziTGu82s4AimfeQ8kA+soGVqvVbIPPKDIA28dtn7b9ddvvTrUeoG9MXe4e\n4xDtSBIAbD9T0p9IOi7pZZLeYvulKdYF9K18MOIire4QaJtJlQFcK+lMRJyNiCckfUzSjYnWBQxK\n6oM/B71DBNpmUgWAyyU9XFg+l923ky47OB8mjE1XB7386vx9pPo88TlN41ii9915VOi6666TJH3+\n85+XdHCVbi6/Yje/r/hYG8t5p0r1/puW8w/wLsvFcsKury8PkI15ed9976OtileWF/trro3lbZ+D\nVG1Vfv983/Ptqdr3qucX36fJcvG9N7XFFK3Xa63X6yTvnWQWkO1XS1pFxPFs+WZJT0XE+wrPiV3W\nnXKWwRjmM9edOdTljKahzJiqo63ZZKln5VQdDHOp2qpq33aZkVR+fvH96i7n710OMnPU5iygVBnA\nFyW92PaVkr4l6U2S3pJoXcDkdX3AO+rMuq8z7zkf+FNIMgYQEf8n6dck/ZOkByT9dUR8LcW6mhhD\n+jiGbcT0HHWgLT9eLt1iHHq9EGyXdW9LfeeizreIDqmEQwkorRRtVb5QMr+vqmxa9Rlt++LKoZdq\nu9RmCWjwAQAEgJQIANX2CQCbtokAkMYorgQG5mTKZY8p79vcEQAmjA9ud6ZcpizuWx99in6cDiWg\nEahbAir/W358LMsptVUCkvqbVtx2W1X9Yt6++9Zmv+M4cSFKQMAATeVMdcrZDC5EBjACZADptJkB\n9KXttqraJzKA4SADANCpqWQ3uBAZwAiQAaRDBvB0bewTGUA6ZAAztc9ZWNVzuVoTQBEZwAiUL8Q5\nStXMlPK/2953aMspNdm2qtkyfSADmBcyAGAAhnDwnzoy1bQIADPGh6s+2m67ttqHIJsWJaCRqVsC\nyr+wq5xSj2GQOKUplCWGWALK7du2q8IPIA2pjYekzRIQAWBkih+Y5XJZ+fOAVQf4o5aHdMAnAOxn\nSgGg+JohtfGQMAYASfXSY0oXzdGG6dHG3SADGJldzpS3nfFXvV/fZ/hjywCGZooZwBDbeSjIAGYu\nInb+gHAmhTL6BHIEgJFbLpdbAwKzKNozpgPntm0dQp/Ytn1jauexowQ0MtvS4zqDZ5SAdt+WofbX\nTWXATW14VPt2UQIqD/i2vd4powQEAGiMAADM3HK57LTsQolnOCgBjQwloHZNoQRUvHgqt08JKOX8\n+6OuoaAEtD9KQEDHhnzWOoRB3aMMuf3mjAxgZMgA6qu6cnqMV/5WKbfX0DKATXP8yQD2RwYA1DCG\nM+W5ITPoFwFgZPjAoEob/WKfCwzbQlDuFyWgCckHAykBVat636mUgIqq/k+Luvx6i6NKQNJhECAY\n7KbNEhABYGL2PVgRAOYZALr62mXbF/xy2tjacogGPwZge2X7nO17s7/jKdaDp6NEhF10dbY9lJ/N\nRLUkGYDtpaTvR8QHtjyHDGAAyADmmQH0Jf9hItQ3+BJQFgAej4hbtjyHADAABAACAMZl8CWgzDtt\n32f7VtsXJVwPBoDSEzA+tTMA2yclXVrx0O9K+oKk72TLvy/psoj45dLro3jQWCwWWiwWtbYF9bWV\nAeTvVff1TTOATT+PWTSXDKBcZqnaZ4zHer3Wer3+wXI202+4JaALVmBfKenOiHh56X5KQAPQJADk\nB90hBIBNrylPP5xDACgjAEzL4EtAti8rLN4k6VSK9aA7VSWePgfzKDnt56gfDsI8pRoE/qikqyWF\npIckvT0izpeeQwYwAMUz4+IZ/S4lnvLZcZcZwK6vaSMDGPvMlbFnMLjQ4GcB7bRiAsAglA/km0pC\n+XPz5bEHgG3Brmrfx4wAMC2DLwFhOqZaahnzGT3QFjKAmTsqAyj+H00pA9iW7ZABYMjIANC6MZ7p\np97mMbYJsA8ygJkrnx2OKQPYts6q5+2bAUylf5IBTAsZAJLhrHd6+D/FJgSAmSsfHKY6ODrng+BU\n/0/RHCUgVBpLCahqm7c9b44lIExLmyWgY228CTAWxYP6nLMCQKIEhIbGfBClNIK5IwCgkfJBdMwB\nAZgbxgBQadcxgPJryuMBxefXXa5SdwygCmMAGBOmgQIAGiMAoFKdUg7lH2BcKAFho6qrhKXdfiSF\nEhCQBiUg9GbIZ/lD3jZgiMgAsFH5LL/4wyh9ZQBH9ZmqH3w5ChkAxoQMAL0Y27x5MgJgOwIAkhjC\nwXdsAQvoGgEASaxWq0EEAQCbEQCwUdMDOGfgwLAxCIxadv2RkV1/djHFIPCu/YtBYIwJg8DAEZpm\nL5SvMAdkAKilOCV0m74ygH36VtW66ZsYqjYzAAIAkiIAAO2iBIRRoiwDDAsZAJIq/3RknYygeP8u\n69v1ueXXFNdN38RQkQEAABojAADATNUOALbfaPurtp+0/YrSYzfb/rrt07avb76ZGKu8bk/9Hhie\n2mMAtn9C0lOSPiTpXRHxpez+l0n6K0k/I+lySZ+V9JKIeKr0esYAZogxAKCZQYwBRMTpiHiw4qEb\nJd0REU9ExFlJZyRdW3c9AIA0UowBvFDSucLyOR1kAgCAATm27UHbJyVdWvHQeyPizj3WQz6N0WC8\nAnOxNQBExM/WeM9vSrqisPyi7L6nKX6VwGKx0GKxqLE6oF18iymGZL1ea71eJ3nvxheC2f6cpN+O\niH/PlvNB4Gt1OAj8Y+URXwaB52kMg8DAkA1iENj2TbYflvRqSZ+2/RlJiogHJH1c0gOSPiPpHRzp\n0ZWI4CAO7IivgkCnUmcAdbcp9TqAtgwiAwC6xMAs0D4yAHSqbgaQsq+QAWBMyAAwGZzZA/0hA0Cn\nyhlA3gfIAIDdkAEAABojAKBTlHyA4SAAoFPFq2wJBkC/GANAL8pfucwYALAbxgCABMhIMDdkAOjF\nEDMA+iPGgAwAaBln/5gjMgD0YmgZAH0RY0EGgNGrOuMu3rdcLrVcLjv5dk/O/jFXZAAYhOIVwvQL\nYDMyAABAYwQAAJgpAgAAzBQBAIPCgCzQHQaBMQgM/gK7YRAYANAYAQAAZooAAAAzRQAAgJkiAADA\nTBEAAGCmCAAAMFMEAAwCF4AB3eNCMAAYkUFcCGb7jba/avtJ268o3H+l7f+xfW/296dtbCgAoF1N\nSkCnJN0k6V8qHjsTEddkf+9osI5ZWK/XfW/CYNAWh2iLQ7RFGrUDQEScjogH29yYuaJzH6ItDtEW\nh2iLNFINAl+VlX/Wtl+baB0AgAaObXvQ9klJl1Y89N6IuHPDy74l6YqIeCwbG/ik7Z+MiO833FYA\nQIsazwKy/TlJ74qIL+3zuG2mAAFADW3NAtqaAezhBxtj+wWSHouIJ23/qKQXS/pG+QVt7QAAoJ4m\n00Bvsv2wpFdL+rTtz2QPXSfpPtv3SvobSW+PiO8131QAQJt6uxAMANCvXr4KwvZx26dtf932u/vY\nhq7ZPmv7/mx21D3ZfRfbPmn7Qdt32b6o8Pybs/Y5bfv6/ra8Odu32T5v+1Thvr333fYrbZ/KHvuj\nrvejDRvaYmX7XOHiyRsKj02yLWxfYftz2cWkX7H969n9s+sXW9oifb+IiE7/JD1T0hlJV0p6lqQv\nS3pp19vRw34/JOni0n3vl/Q72e13S/qD7PbLsnZ5VtZOZyQ9o+99aLDvr5N0jaRTNfc9z1TvkXRt\ndvsfJB3ve99aaoulpN+qeO5k20IHswuvzm4/T9J/SHrpHPvFlrZI3i/6yACu1cGVwmcj4glJH5N0\nYw/b0YfywPfrJd2e3b5d0huy2zdKuiMinoiIszr4D762ky1MICLulvRY6e599v1Vti+T9PyIuCd7\n3kcLrxmNDW0hPb1vSBNui4j4dkR8Obv9uKSvSbpcM+wXW9pCStwv+ggAl0t6uLB8Toc7O2Uh6bO2\nv2j7V7L7LomI89nt85IuyW6/UAftkptiG+277+X7v6lptck7bd9n+9ZC2WMWbWH7Sh1kRf+mmfeL\nQlt8Ibsrab/oIwDMddT5NRFxjaQbJP2q7dcVH4yDnG1b20y23XbY96n7oKSrJF0t6RFJt/S7Od2x\n/TxJfyfpN6J0sejc+kXWFn+rg7Z4XB30iz4CwDclXVFYvkIXRq1JiohHsn+/I+kTOijpnLd9qSRl\n6duj2dPLbfSi7L4p2Wffz2X3v6h0/yTaJCIejYykD+uw3DfptrD9LB0c/P88Ij6Z3T3LflFoi7/I\n26KLftFHAPiipBf74Gujny3pTZI+1cN2dMb2c2w/P7v9XEnX6+DbVD8l6a3Z094qKf8QfErSm20/\n2/ZVOriY7h5Ny177HhHflvTftl9l25J+ofCaUcsOdLmbdNA3pAm3Rbbdt0p6ICL+sPDQ7PrFprbo\npF/0NOp9gw5Gus9IurmPbeh4f6/Swaj9lyV9Jd9nSRdL+qykByXdJemiwmvem7XPaUk/1/c+NNz/\nO3TwHVH/q4Pxn1+qs++SXpl9CM5I+uO+96ultnibDgbr7pd0X/aBvWTqbSHptZKeyj4T92Z/x+fY\nLza0xQ1d9AsuBAOAmeI3gQFgpggAADBTBAAAmCkCAADMFAEAAGaKAAAAM0UAAICZIgAAwEz9P3mP\naG7sLdaqAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x104a81ed0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# READ DNA SEQUENCE FROM FASTA FILE\n",
    "from Bio import SeqIO\n",
    "with open(\"../data/human_beta_globin.fasta\") as f:\n",
    "    for record in SeqIO.parse(f, \"fasta\"):\n",
    "        seq = record # there is only 1 record.\n",
    "seq = str( seq.seq )\n",
    "\n",
    "# INITIALIZE MATPLOTLIB FUNCTIONALITY\n",
    "import matplotlib\n",
    "%pylab inline\n",
    "print \"(Ignore any warnings...)\"\n",
    "\n",
    "# PREDICTION PARAMETERS\n",
    "window = 120\n",
    "overlap = 10\n",
    "\n",
    "# COMPUTE AND PLOT PREDICTION\n",
    "x = range(0, len(seq), overlap)\n",
    "y = exon_predictor(seq, window, overlap)\n",
    "bar ( x,y )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hopefully you will see a plot like this\n",
    "\n",
    "<img src=\"files/images/week3_exonintron.png\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">If you get an error when running the code above you might not have ```matplotlib``` installed. On mac or linux run in a terminal: \n",
    "\n",
    "><cb>sudo pip install matplotlib</cb>\n",
    "\n",
    ">Windows users should already have matplotlib installed from the pythonxy distribution.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare the predicted exons with experimentally validated data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On this plot, it is easy to identify the three regions whose coding potential is positive. Check that those regions are the three real exons that were experimentally verified here: \n",
    "\n",
    "http://www.ebi.ac.uk/ena/data/view/V00499\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The three validated exon regions are:"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "exon1 : \n",
    "exon2 : \n",
    "exon3 : "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. [Optional] Conditional Expressions (Advanced)\n",
    "\n",
    "Python can make some rather nice *conditional expressions* in a single line using the syntax\n",
    "\n",
    "    expression1 if condition else expression2\n",
    "\n",
    "An example is:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATCG\n"
     ]
    }
   ],
   "source": [
    "type = \"DNA\"\n",
    "bases = \"ATCG\" if type==\"DNA\" else \"AUCG\"\n",
    "print bases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "make a one-line command that can print the average of a list using\n",
    "\n",
    "    sum(A)/len(A)\n",
    "\n",
    "However, if the list is empty it should print \"N/A\". Test it on these two lists:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "A = [1,2,3,4]\n",
    "B = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "N/A\n"
     ]
    }
   ],
   "source": [
    "# Make one-liner here and try it with A and B.\n",
    "print sum(A)/len(A) if len(A)>0 else \"N/A\"\n",
    "print sum(B)/len(B) if len(B)>0 else \"N/A\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"../pix/FinishFlag.jpg\">\n",
    "<br><br><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "\n",
    "\n",
    "def css_styling():\n",
    "    styles = open(\"../styles/custom.css\", \"r\").read()\n",
    "    return HTML(styles)\n",
    "css_styling()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
